File: lecture-notes/games-dev-pathway/week-02-design-patterns/code-examples/singleton.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <string>
#include <format>

struct Card
{
    std::string suit;
    int value;
    std::string toString() const
    {
        std::string name;
        if (value == 11)
            name = "Jack";
        else if (value == 12)
            name = "Queen";
        else if (value == 13)
            name = "King";
        else if (value == 14)
            name = "Ace";
        else
            name = std::to_string(value);
        return name + " of " + suit;
    }
};

class GameManager
{
public:
    static GameManager &getInstance()
    {
        static GameManager instance;
        return instance;
    }

    GameManager(const GameManager &) = delete;
    GameManager &operator=(const GameManager &) = delete;

    void startGame(const std::string &p1Name, const std::string &p2Name)
    {
        if (running)
        {
            std::cout << "A game is already in progress\n";
            return;
        }
        running = true;
        gamesPlayed++;
        std::cout << std::format("── Game {} starting: {} vs {} ──\n",
                                 gamesPlayed, p1Name, p2Name);
        runGame(p1Name, p2Name);
        running = false;
    }

    int getGamesPlayed() const { return gamesPlayed; }

private:
    GameManager() = default;

    bool running = false;
    int gamesPlayed = 0;

    void runGame(const std::string &p1Name, const std::string &p2Name)
    {
        std::vector<Card> deck;
        for (const auto &suit : {"Spades", "Hearts", "Clubs", "Diamonds"})
            for (int v = 2; v <= 14; ++v)
                deck.push_back({suit, v});
        std::mt19937 rng(std::random_device{}());
        std::shuffle(deck.begin(), deck.end(), rng);

        std::vector<Card> hand1(deck.begin(), deck.begin() + 26);
        std::vector<Card> hand2(deck.begin() + 26, deck.end());

        int w1 = 0, w2 = 0;
        for (int round = 1; round <= 5; ++round)
        {
            Card c1 = hand1.back();
            hand1.pop_back();
            Card c2 = hand2.back();
            hand2.pop_back();

            std::cout << std::format("  Round {}: {} [{}] vs {} [{}]  =>  ",
                                     round, p1Name, c1.toString(), p2Name, c2.toString());

            if (c1.value > c2.value)
            {
                ++w1;
                std::cout << p1Name << " wins\n";
            }
            else if (c2.value > c1.value)
            {
                ++w2;
                std::cout << p2Name << " wins\n";
            }
            else
            {
                std::cout << "Tie\n";
            }
        }

        std::cout << std::format("Result => {} wins: {}  {} wins: {}\n\n",
                                 p1Name, w1, p2Name, w2);
    }
};

int main()
{

    GameManager::getInstance().startGame("Alice", "Bob");
    GameManager::getInstance().startGame("Carol", "Dave");
    GameManager::getInstance().startGame("Alice", "Bob");

    std::cout << std::format("Total games played: {}\n",
                             GameManager::getInstance().getGamesPlayed());
    return 0;
}

File: lecture-notes/app-dev-pathway/week-02-design-patterns/code-examples/singleton.py
import random


class Card:
    def __init__(self, suit, value):
        self.suit = suit
        self.value = value

    def __str__(self):
        names = {11: "Jack", 12: "Queen", 13: "King", 14: "Ace"}
        name = names.get(self.value, str(self.value))
        return f"{name} of {self.suit}"


class GameManager:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.running = False
            cls._instance.games_played = 0
        return cls._instance

    @classmethod
    def get_instance(cls):
        return cls()

    def start_game(self, p1_name, p2_name):
        if self.running:
            print("A game is already in progress!")
            return

        self.running = True
        self.games_played += 1
        print(f"── Game {self.games_played} starting: {p1_name} vs {p2_name} ──")
        self._run_game(p1_name, p2_name)
        self.running = False

    def get_games_played(self):
        return self.games_played

    def _run_game(self, p1_name, p2_name):
        suits = ["Spades", "Hearts", "Clubs", "Diamonds"]
        deck = [Card(suit, value) for suit in suits for value in range(2, 15)]
        random.shuffle(deck)

        hand1 = deck[:26]
        hand2 = deck[26:]

        w1 = 0
        w2 = 0

        for round_num in range(1, 6):
            c1 = hand1.pop()
            c2 = hand2.pop()

            print(f"  Round {round_num}: {p1_name} [{c1}] vs {p2_name} [{c2}]  =>  ", end="")

            if c1.value > c2.value:
                w1 += 1
                print(f"{p1_name} wins")
            elif c2.value > c1.value:
                w2 += 1
                print(f"{p2_name} wins")
            else:
                print("Tie")

        print(f"Result => {p1_name} wins: {w1}  {p2_name} wins: {w2}\n")


GameManager.get_instance().start_game("Alice", "Bob")
GameManager.get_instance().start_game("Carol", "Dave")

print(f"Total games played: {GameManager.get_instance().get_games_played()}")